src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1251 @@
+	 */
+	LOCKTAG tag = { 0 };
+	SET_LOCKTAG_CITUS_OPERATION(tag, CITUS_BACKGROUND_TASK_MONITOR);
+	const bool sessionLock = true;
+	const bool dontWait = true;
+	LockAcquireResult locked =
+		LockAcquire(&tag, AccessExclusiveLock, sessionLock, dontWait);
+	if (locked == LOCKACQUIRE_NOT_AVAIL)
+	{
+		ereport(ERROR, (errmsg("background task queue monitor already running for "
+							   "database")));
+		exit(0);
+	}
+
+	/* make worker recognizable in pg_stat_activity */
+	pgstat_report_appname("citus background task queue monitor");
+
+	ereport(DEBUG1, (errmsg("started citus background task queue monitor")));
+
+	/* TODO this is here for debugging purposses, remove before merge. */
+	if (BackgroundTaskMonitorDebugDelay)
+	{
+		pg_usleep(30 * 1000 * 1000);
+	}
+
+	MemoryContext perTaskContext = AllocSetContextCreate(CurrentMemoryContext,
+														 "PerTaskContext",
+														 ALLOCSET_DEFAULT_MINSIZE,
+														 ALLOCSET_DEFAULT_INITSIZE,
+														 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * First we find all jobs that are running, we need to check if they are still running
+	 * if not reset their state back to scheduled.
+	 */
+	{
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		ResetRunningBackgroundTasks();
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+
+	MemoryContext oldContextPerJob = MemoryContextSwitchTo(perTaskContext);
+	bool hasJobs = true;
+	while (hasJobs)
+	{
+		MemoryContextReset(perTaskContext);
+
+		CHECK_FOR_INTERRUPTS();
+
+		InvalidateMetadataSystemCache();
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		/*
+		 * We need to load the task into the perTaskContext as we will switch contexts
+		 * later due to the committing and starting of new transactions
+		 */
+		MemoryContext oldContext = MemoryContextSwitchTo(perTaskContext);
+		BackgroundTask *task = GetRunnableBackgroundTask();
+
+		if (!task)
+		{
+			MemoryContextSwitchTo(oldContext);
+
+			PopActiveSnapshot();
+			CommitTransactionCommand();
+
+			hasJobs = false;
+			break;
+		}
+
+		/* we load the database name and username here as we are still in a transaction */
+		char *databaseName = get_database_name(MyDatabaseId);
+		char *userName = GetUserNameFromId(task->owner, false);
+
+		MemoryContextSwitchTo(oldContext);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		/*
+		 * The background worker needs to be started outside of the transaction, otherwise
+		 * it will complain about leaking shared memory segments used, among other things,
+		 * to communicate the output of the backend.
+		 */
+		dsm_segment *seg = NULL;
+		BackgroundWorkerHandle *handle =
+			StartCitusBackgroundJobExecuter(databaseName, userName, task->command,
+											task->taskid, &seg);
+
+		if (handle == NULL)
+		{
+			ereport(ERROR, (errmsg("unable to start background worker for background "
+								   "task execution")));
```

Comment:
great catch, haven't paid attention about that case.

what could be a reason this backend wouldn't start? mainly exhaust of allowed background processes I guess.
Maybe we should have a loop here that keeps retrying?

Reply:
yes, reaching max_worker_processes would be the typical scenario. 

> Maybe we should have a loop here that keeps retrying?

yeh, with plenty of wait time

Reply:
I would reuse the existing loop instead, by stopping the transaction, sleeping, and running `continue`. Otherwise you might keep the snapshot open for a very long time.

Reply:
There is no transaction and snapshot open at this time I believe.

The background worker needs to specifically be started outside of a transaction. Otherwise the transaction gives a lot of errors when closing due to a leaking background worker.

as per comment above:
```C
		/*
		 * The background worker needs to be started outside of the transaction, otherwise
		 * it will complain about leaking shared memory segments used, among other things,
		 * to communicate the output of the backend.
		 */
```

Sleeping and continuing is probably the way to go.

Question would be on how long to sleep though.

Initially I would suggest the background task queue interval, but there is an orthogonal discussion on lowering that delay to give a more instant feedback loop for the user starting a rebalance.

I don't think there is a way if being interrupted by postmaster when a background worker slot becomes available, is there?